### ISSUE TITLE
false alarm for (su compress) options used in tandem

### ISSUE THREAD

**Issue by bazzinotti** [logrotate/logrotate#169]
I found a bug in logrotate. I also provide a patch to fix it.

__Summary__: stderr output from `switch_user_permanently` is sent to `compressLogFile` pipe. `compressLogFile` thinks it is reading from the compresscmd stderr. It incorrectly believes in this case that the compresscmd incurred an error, when in fact it was just `switch_user_permanently` providing a status report.

Here is an excerpt:

```
# logrotate /etc/logrotate.conf -v
...
compressing log with: /usr/bin/xz
error: Compressing program wrote following message to stderr when compressing log /var/log/bazz/offlineimap.log-20180205:
switching uid to 501 and gid to 1000
switching euid to 0 and egid to 0
...
```

The follow excerpts I've made as I researched the bug.

__switch_user_permanently__
https://github.com/logrotate/logrotate/blob/a4f61c1b470fb3868ca7bfe37b1e36ad3db246d9/logrotate.c#L178-L179

Proof that MESS_DEBUG will cause stderr output by default.

https://github.com/logrotate/logrotate/blob/a4f61c1b470fb3868ca7bfe37b1e36ad3db246d9/log.h#L7-L7
https://github.com/logrotate/logrotate/blob/a4f61c1b470fb3868ca7bfe37b1e36ad3db246d9/log.c#L13-L13
https://github.com/logrotate/logrotate/blob/a4f61c1b470fb3868ca7bfe37b1e36ad3db246d9/log.c#L58-L66

__compressLogFile__
https://github.com/logrotate/logrotate/blob/a4f61c1b470fb3868ca7bfe37b1e36ad3db246d9/logrotate.c#L763
https://github.com/logrotate/logrotate/blob/a4f61c1b470fb3868ca7bfe37b1e36ad3db246d9/logrotate.c#L748-L787

The call to `switch_user_permanently` on L765 was not anticipated to write to stderr. The fix for this is simple. Here's a patch:

```
diff -rupN logrotate-3.13.0/logrotate.c 2/logrotate.c
--- logrotate-3.13.0/logrotate.c	2017-10-12 11:19:41.000000000 -0400
+++ 2/logrotate.c	2018-02-06 01:18:01.124995758 -0500
@@ -760,11 +760,12 @@ static int compressLogFile(char *name, s
 
 	movefd(inFile, STDIN_FILENO);
 	movefd(outFile, STDOUT_FILENO);
-	movefd(compressPipe[1], STDERR_FILENO);
 
 	if (switch_user_permanently(log) != 0) {
 		exit(1);
 	}
+
+	movefd(compressPipe[1], STDERR_FILENO);
 
 	envInFilename = alloca(strlen("LOGROTATE_COMPRESSED_FILENAME=") + strlen(name) + 2);
 	sprintf(envInFilename, "LOGROTATE_COMPRESSED_FILENAME=%s", name);
```

All this does is postpone moving the pipe to STDERR_FILENO until after the `switch_user_permanently` call. The `switch_user_permanently`'s prints will go directly to stderr now.

I tested this with `logrotate /etc/logrotate.d/bazz -v` which is the following file
```
# Copyright 2018 Michael Bazzinotti <bazz@bazz1.com>

/var/log/bazz/*.log {
  copytruncate
  compress
  compresscmd /usr/bin/xz
  compressoptions -9
  uncompresscmd /usr/bin/unxz
  dateyesterday
  nocreate
  notifempty
  maxsize 4M
  rotate 16
  weekly
  missingok
  su bazz bazz
}
```

**Comment by bazzinotti** [comment]
With the patch in place, here is following related output of `logrotate /etc/logrotate.d/bazz -v`

```
compressing log with: /usr/bin/xz
switching uid to 501 and gid to 1000
switching euid to 0 and egid to 0
```

Just to be sure everything works when the compresscmd does indeed incur an error, I tried using an invalid option to xz, `-b` and got the following response.
```
...
compressing log with: /usr/bin/xz
switching uid to 501 and gid to 1000
error: Compressing program wrote following message to stderr when compressing log /var/log/bazz/offlineimap.log.1:
/usr/bin/xz: invalid option -- 'b'
/usr/bin/xz: Try `/usr/bin/xz --help' for more information.
error: failed to compress log /var/log/bazz/offlineimap.log.1
switching euid to 0 and egid to 0
...
```

In inclusion the patch is successful :smiley:

**Comment by kdudka** [comment]
Indeed.  Thanks for the detailed bug report and the patch!

**Comment by lsh-0** [comment]
Thanks guys for this patch and for logrotate in general.

Is it possible to get this fix as a patch release in `3.11` ? We're upgrading our machines to Ubuntu 18.04 LTS and we're getting hit by this false warning about 1-5 times per machine across 20 machines (so far), every day. We'll be on 18.04 for several years at least.

**Comment by kdudka** [comment]
Yes, it is really easy technically-wise.  The patch would look like this:
```
--- a/logrotate.c
+++ b/logrotate.c
@@ -712,20 +712,21 @@ static int compressLogFile(char *name, struct logInfo *log, struct stat *sb)

     if (!FORK_OR_VFORK()) {
        dup2(inFile, 0);
        close(inFile);
        dup2(outFile, 1);
        close(outFile);
-       dup2(compressPipe[1], 2);
        close(compressPipe[0]);
-       close(compressPipe[1]);

        if (switch_user_permanently(log) != 0) {
                DOEXIT(1);
        }

+       dup2(compressPipe[1], 2);
+       close(compressPipe[1]);
+
        envInFilename = alloca(strlen("LOGROTATE_COMPRESSED_FILENAME=") + strlen(name) + 2);
        sprintf(envInFilename, "LOGROTATE_COMPRESSED_FILENAME=%s", name);
        putenv(envInFilename);
        execvp(fullCommand[0], (void *) fullCommand);
        DOEXIT(1);
     }
```

I cannot speak for Ubuntu folks though.  It is ultimately their decision whether they will release an official update of logrotate with the above patch included.
